lib/charms/opensearch/v0/opensearch_snapshots.py

```diff
@@ -0,0 +1,1628 @@
+        # handle case where this was deferred in the above case, then the s3 relation was severed
+        if object_storage_type != "s3" or not cfg or not creds:
+            logger.warning("No S3 object storage configuration.")
+            if self.charm.unit.is_leader():
+                self.charm.status.set(BlockedStatus(BackupRelDataIncomplete), app=True)
+                self.charm.status.set(BlockedStatus(BackupRelDataIncomplete))
+            return
+
+        if self.charm.unit.is_leader():
+            self.charm.status.clear(BackupRelDataIncomplete, app=True)
+            self.charm.status.clear(BackupRelDataIncomplete)
+
+        # apply locally (leader does cluster-level config)
+        self.charm.keystore_manager.put_entries(
+            {
+                "s3.client.default.access_key": cfg.s3.credentials.access_key,
+                "s3.client.default.secret_key": cfg.s3.credentials.secret_key,
+            }
+        )
+        logger.info("S3 credentials are added to keystore.")
+        self.charm.keystore_manager.reload()
+
+        need_custom_ca = self.charm.snapshots_manager.requires_custom_s3_ca(
+            object_storage_type, cfg
+        )
+        new_chain = cfg.s3.tls_ca_chain if need_custom_ca else None
+
+        if new_chain:
+            self.charm.snapshots_manager.store_s3_ca(new_chain)
+            logger.info("S3 CA stored.")
+        else:
+            if self.charm.snapshots_manager.is_custom_s3_ca_stored():
+                self.charm.snapshots_manager.remove_s3_ca()
+                logger.info("S3 CA removed.")
+
+        self._restart_for_ca(new_chain, reason="apply object storage CA change")
+
+        try:
+            self.charm.snapshots_manager.ensure_repository(object_storage_type, cfg)
+            self.charm.snapshots_manager.verify_repository("s3")
+        except OpenSearchHttpError as e:
+            if self.charm.unit.is_leader():
+                self._set_app_to_blocked(self._classify_os_repo_errors(e))
+            return
+
+        if self.charm.unit.is_leader():
+            self.charm.status.clear(BackupCredentialKeysIncorrect, app=True)
+            self.charm.status.clear(BackupCredentialKeysIncorrect)
+            self.charm.status.clear(BackupCredentialCAIncorrect, app=True)
+            self.charm.status.clear(BackupCredentialCAIncorrect)
+
+        self.charm.peer_cluster_provider.refresh_relation_data(event, can_defer=False)
+        self._broadcast_storage_trigger(kind="s3", action="changed")
+
+    def _on_s3_credentials_gone(self, event: CredentialsGoneEvent) -> None:
+        """Handler for s3 credentials gone event."""
+        if self.charm.unit.is_leader():
+            self.charm.status.clear(BackupRelShouldNotExist, app=True)
+            self.charm.status.clear(BackupRelShouldNotExist)
+
+        if self.charm.snapshots_manager.resolver.get_storage_type() == ObjectStorageType.CONFLICT:
+            return
```

Comment:
we are returning here but are not cleaning up? 
assume the case where: first relation is s3 => charm configured, added azure relation => conflict, we remove s3 => needs to be cleaned up

I still think we need to store the current configured object storage in the relation (Gothenburg discussion)